src/accelerate/data_loader.py

```diff
@@ -606,7 +606,16 @@ def _fetch_batches(self, iterator):
                     batches = []
                     for _ in range(self.state.num_processes):
                         batches.append(next(iterator))
-                    batch = concatenate(batches, dim=0)
+                    try:
+                        batch = concatenate(batches, dim=0)
+                    except RuntimeError:
+                        raise RuntimeError(
+                            "You can't use batches of different size with `dispatch_batches=True` or when using an `IterableDataset`."
```

Comment:
Alternative: we can directly check if a user has an `IterableDataset` and raise the right part of the err